R/distribution.R

```diff
@@ -0,0 +1,203 @@
+      not_implemented_error()
+    },
+    
+    #'  Returns entropy of distribution, batched over batch_shape.
+    #'  @return Tensor of shape batch_shape.
+    entropy = function() {
+      not_implemented_error()
+    },
+    
+    #'  Returns perplexity of distribution, batched over batch_shape.
+    #'  @param Tensor of shape batch_shape.
+    perplexity = function() {
+      not_implemented_error()
+    },
+    
+    #' Returns the size of the sample returned by the distribution, given
+    #' a `sample_shape`. Note, that the batch and event shapes of a distribution
+    #' instance are fixed at the time of construction. If this is empty, the
+    #' returned shape is upcast to (1,).
+    #' @param sample_shape (torch_Size): the size of the sample to be drawn.
+    `_extended_shape` = function(sample_shape = NULL){
+     # if (!inherits(sample_shape, "torch_size"))
+     #   sample_shape <- torch_size(sample_shape) #!
+     sample_shape + self$batch_shape + self$event_shape
+    },
+    
+    #' Argument validation for distribution methods such as `log_prob`,
+    #' `cdf` and `icdf`. The rightmost dimensions of a value to be
+    #' scored via these methods must agree with the distribution's batch
+    #' and event shapes.
+    #' @param value (Tensor): the tensor whose log probability is to be
+    #' computed by the `log_prob` method.
+    `_validate_sample` = function(value){
+      
+      if (!inherits(value, "torch_Tensor"))
+        value_error('The value argument to log_prob must be a Tensor')
+      
+      event_dim_start <-length(value$size()) - length(self$`_event_shape`)
+      
+      if (value$size()[event_dim_start, ] != self$`_event_shape`)
+        value_error('The right-most size of value must match event_shape: 
+                    {value$size()} vs {self$`_event_shape`}.')
+      
+      actual_shape <- value$size()
+      expected_shape <- self$`_batch_shape` + self$`_event_shape`
+      
+      shape_length <- length(actual_shape)
+      
+      for (idx in shape_length:1) {
+        i <- actual_shape[idx]
+        j <- expected_shape[idx]
+        
+        if (i != 1 && j != 1 && i != j)
+          value_error('Value is not broadcastable with 
+                      batch_shape+event_shape: {actual_shape} vs {expected_shape}.')
+      }
+      
+      if (!self$support$check(value)$all())
+        value_error('The value argument must be within the support')
+    },
+    
+    `_get_checked_instance` = function(cls, .instance = NULL){
+      if (is.null(.instance) && self$initialize == cls$initialize)
+        not_implemented_error("Subclass {class(self)} of {class(.instance)} ",
+                              "that defines a custom __init__ method ",
+                              "must also define a custom `_expand()` method.")
+      
+      # TODO: better mechanism to get own instance's class
+      if (is.null(.instance))
+        return(eval(parse(text = glue("torch:::{class(self)}")))$new())
```

Comment:
Yes something like this: https://github.com/r-lib/R6/issues/144#issuecomment-523075721 or a `distribution` helper to reate the class that saves the generator in an attribute of the instance. We did something similar  for `nn_modules` and datasets.